File: oolab/src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;

public class Animal implements WorldElement{
    private MapDirection orientation;
    private Vector2d position;

    public Animal() {
        orientation = MapDirection.NORTH;
        position = new Vector2d(2, 2);
    }

    public Animal(Vector2d position) {
        this();
        this.position = position;
    }

    public MapDirection getOrientation() {
        return orientation;
    }

    public Vector2d getPosition() {
        return position;
    }

//    @Override
//    public String toString() {
//        return "Pozycja: " + position.toString() + ", orientacja: " + orientation.toString();
//    }


    @Override
    public String toString() {
        return orientation.toStringShort();
    }

    public boolean isAt(Vector2d position) {
        return this.position.equals(position);
    }

    public Vector2d move(MoveDirection direction, MoveValidator validator) {
        switch (direction) {
            case FORWARD -> {
                Vector2d temp = position.add(orientation.toUnitVector());
                if (validator.canMoveTo(temp))
                    position = temp;
            }
            case BACKWARD -> {
                Vector2d temp = position.subtract(orientation.toUnitVector());
                if (validator.canMoveTo(temp))
                    position = temp;
            }
            case LEFT -> orientation = orientation.previous();
            case RIGHT -> orientation = orientation.next();

        }
        return position;
    }

}

File: OOKotlin/src/main/kotlin/Animal.kt
import Vector2d.Companion.toUnitVector

class Animal(private var position: Vector2d = Vector2d(2, 2)) {
    private var orientation: MapDirection = MapDirection.NORTH

    fun getOrientation() = orientation

    fun getPosition() = position

    override fun toString(): String = orientation.toStringShort()

    fun isAt(position: Vector2d) = this.position == position


    fun move(direction: MoveDirection, validator: IMoveValidator): Vector2d {
        when (direction) {
            MoveDirection.FORWARD -> {
                val temp = position + orientation.toUnitVector()
                if (validator.canMoveTo(temp)) position = temp
            }

            MoveDirection.BACKWARD -> {
                val temp: Vector2d = position - orientation.toUnitVector()
                if (validator.canMoveTo(temp)) position = temp
            }

            MoveDirection.LEFT -> orientation = orientation.previous()
            MoveDirection.RIGHT -> orientation = orientation.next()
        }
        return position
    }
}
